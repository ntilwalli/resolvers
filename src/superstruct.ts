import {
  appendErrors,
  transformToNestObject,
  Resolver,
  FieldValues,
} from 'react-hook-form';
import { validate, StructFailure, StructError } from 'superstruct';
import convertArrayToPathName from './utils/convertArrayToPathName';

function getParseReducer(validateAllFieldCriteria: boolean) {
  return function blah(
    previous: Record<string, any>,
    { path, message = '', type }: StructFailure,
  ) {
    const currentPath = convertArrayToPathName(path);

    return {
      ...previous,
      ...(path
        ? previous[currentPath] && validateAllFieldCriteria
          ? {
              [currentPath]: appendErrors(
                currentPath,
                validateAllFieldCriteria,
                previous,
                type || '',
                message,
              ),
            }
          : {
              [currentPath]: previous[currentPath] || {
                message,
                type,
                ...(validateAllFieldCriteria
                  ? {
                      types: { [type || '']: message || true },
                    }
                  : {}),
              },
            }
        : {}),
    };
  };
}

const parseErrorSchema = (
  error: StructError,
  validateAllFieldCriteria: boolean,
) => {
  const failures: Iterable<StructFailure> = error.failures();
  const reified: StructFailure[] = Array.from(failures);
  return reified.length
    ? reified.reduce(getParseReducer(validateAllFieldCriteria), {})
    : [];
};

export const superstructResolver = <TFieldValues extends FieldValues>(
  schema: any,
): Resolver<TFieldValues> => async (
  values,
  _,
  validateAllFieldCriteria = true,
) => {
  try {
    const [error] = validate(values, schema);
    if (!error) {
      return {
        values: values,
        errors: {},
      };
    } else {
      throw error;
    }
  } catch (e) {
    const out = parseErrorSchema(e, validateAllFieldCriteria);
    return {
      values: {},
      errors: transformToNestObject(out),
    };
  }
};
