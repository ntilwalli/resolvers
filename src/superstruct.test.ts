import { object, number, string, boolean, optional, array } from 'superstruct';
import { superstructResolver } from './superstruct';

const Article = object({
  id: number(),
  title: string(),
  isPublished: optional(boolean()),
  tags: array(string()),
  author: object({
    id: number(),
  }),
});

describe('superstructResolver', () => {
  it('should return correct value', async () => {
    const data = {
      id: 2,
      title: 'test',
      tags: ['news', 'features'],
      author: {
        id: 1,
      },
    };
    expect(await superstructResolver(Article)(data)).toEqual({
      values: data,
      errors: {},
    });
  });

  it('should return errors', async () => {
    const data = {
      id: '2',
      title: 2,
      tags: [2, 3],
      author: {
        id: 'test',
      },
    };

    expect(await superstructResolver(Article)(data)).toMatchSnapshot();
  });
});
