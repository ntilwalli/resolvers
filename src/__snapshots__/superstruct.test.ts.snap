// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`superstructResolver should return errors 1`] = `
Object {
  "errors": Object {
    "author": Object {
      "id": Object {
        "message": "",
        "type": "number",
        "types": Object {
          "number": true,
        },
      },
    },
    "id": Object {
      "message": "",
      "type": "number",
      "types": Object {
        "number": true,
      },
    },
    "tags": Array [
      Object {
        "message": "",
        "type": "string",
        "types": Object {
          "string": true,
        },
      },
      Object {
        "message": "",
        "type": "string",
        "types": Object {
          "string": true,
        },
      },
    ],
    "title": Object {
      "message": "",
      "type": "string",
      "types": Object {
        "string": true,
      },
    },
  },
  "values": Object {},
}
`;
